nominatim/tokenizer/sanitizers/delete_tags.py

```diff
@@ -0,0 +1,146 @@
+                  to learn more about address rank.
+
+
+"""
+from typing import Callable, List, Optional, Pattern, Tuple, Sequence
+import re
+
+from nominatim.tokenizer.sanitizers.base import ProcessInfo
+from nominatim.data.place_name import PlaceName
+from nominatim.tokenizer.sanitizers.config import SanitizerConfig
+
+class _TagSanitizer:
+
+    def __init__(self, config: SanitizerConfig) -> None:
+        self.type = config.get('type', 'name')
+        self.filter_kind = config.get_filter_kind()
+        self.country_codes = config.get_string_list('country_code', [])
+        self.rank_address_intervals = self._compute_rank_intervals( \
+                                            config.get_string_list('rank_address', ['0-30']))
+
+        suffixregexps = config.get_string_list('suffix', [])
+        self.suffix_regexp = [re.compile(r) for r in suffixregexps]
+
+        nameregexps = config.get_string_list('name', [])
+        self.name_regexp = [re.compile(r) for r in nameregexps]
+
+
+
+    def __call__(self, obj: ProcessInfo) -> None:
+        tags = obj.names if self.type == 'name' else obj.address
+
+        if not tags:
+            return
+
+        filtered_tags: List[PlaceName] = []
+
+        for tag in tags:
+            keep_tag = True
+
+            if self.filter_kind(tag.kind):
+
+                if (obj.place.country_code in self.country_codes and
+                    self._in_rank_addresses(obj.place.rank_address)):
+
+                    if self._matches(tag.suffix, self.suffix_regexp) and \
+                        self._matches(tag.name, self.name_regexp):
+                        keep_tag = False
+
+            if keep_tag:
+                filtered_tags.append(tag)
+
+
+        if self.type == 'name':
+            obj.names = filtered_tags
+        else:
+            obj.address = filtered_tags
+
+
+    def _compute_rank_intervals(self, ranks: Sequence[str]) -> List[Tuple[int, int]]:
+        """ Creates a list of rank address intervals from the list of rank address
+            provided in the sanitizer configuration. Each interval is of the type
+            (from, to) which tells the range of rank addresses that lie in it.
+
+            Note that single ranks(r) provided in the rank address configuration are also
+            converted into an interval of the form (r, r).
+        """
+        intervals: List[Tuple[int, int]] = []
+
+        for rank in ranks:
+            intv = [int(x) for x in rank.split('-')]
+
+            intervals.append((intv[0], intv[0]) if len(intv) == 1 \
+                            else (intv[0], intv[1]))
+
+        return intervals
+
+
+    def _in_rank_addresses(self, rank: int) -> bool:
+        """ Return True if the given rank address lies in any of the
+            rank address intervals. Otherwise, returns False.
+         """
+        return any(l <= rank <= r for l, r in self.rank_address_intervals)

```

Comment:
You can do the check for rank more efficiently here. You just need to do a bit more preprocessing: we know that there are only ranks 0-30. So during preprocessing of the incoming configuration you can create a set of allowed rank numbers. That means spell out every single number to be accepted. For example, from the default configuration of '0-30' you can create a `allowed_ranks = set(range(0, 31))` and then simply check `obj.place.rank_address in allowed_ranks`.

(Even more advanced and faster: An alternative to a set would be a tuple of 31 booleans, so that eventually the lookup is `allowed_ranks[obj.place.rank_address]`.)